File: src/boj/dp1/BOJ1932.java
package boj.dp1;

import java.io.*;
import java.util.StringTokenizer;

public class BOJ1932 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int[][] tri = new int[n][n];
        StringTokenizer st;

        tri[0][0] = Integer.parseInt(br.readLine());
        for (int i = 1; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j <= i; j++) {
                if (j == 0 ) {
                    tri[i][j] = tri[i-1][j] + Integer.parseInt(st.nextToken());
                } else if (j == i) {
                    tri[i][j] = tri[i-1][i-1] + Integer.parseInt(st.nextToken());
                } else {
                    tri[i][j] = Math.max(tri[i-1][j-1], tri[i-1][j]) + Integer.parseInt(st.nextToken());
                }
            }
        }
        int max = tri[n-1][0];
        for (int i = 1; i < n ; i++) {
            max = Math.max(max, tri[n-1][i]);
        }
        System.out.println(max);
    }
}

File: ps_kotlin/src/main/kotlin/boj/boj1932.kt
package boj

import java.util.*

fun main() {
    val br = System.`in`.bufferedReader()

    val n = br.readLine().toInt()
    val dp = Array(n) { IntArray(n) }
    dp[0][0] = br.readLine().toInt()

    for (i in 1 until n) {
        val st = StringTokenizer(br.readLine(), " ")
        for (j in 0..i) {
            dp[i][j] = st.nextToken().toInt()
            when (j) {
                0 -> dp[i][j] += dp[i - 1][j]
                i -> dp[i][j] += dp[i - 1][i - 1]
                else -> dp[i][j] += maxOf(dp[i - 1][j - 1], dp[i - 1][j])
            }
        }
    }
    var result = 0
    for (i in 0 until n) {
        result = maxOf(result, dp[n - 1][i])
    }
    println(result)
}